### PR TITLE
#163383200 Users should see their reading stats

### PR DESCRIPTION
**Pull request by codjoero**
#### What does this PR do?
This PR enables a user to be able to see their reading stats
This task should also allow an author to see the reading stats for each of their article

#### Description of Task to be completed?
* add a model for a user'reading stats
* add functionality for when a user reads article.
* add a read_stats field to handle the data for the number of times an article has been read.
* add view functionality to handle the endpoint that retrieves the read stats.

#### How should this be manually tested?
* Visit the [PR's](https://ah-backend-summer-stagin-pr-45.herokuapp.com) app with postman.

#### Screenshots
Get read stats for a user `/api/v1/profiles/user/readstats`
image

Reading an article `/api/v1/articles/<slug>`
image

Author get their article `/api/v1/articles/<slug>`
image

### Related story
[163383200](https://www.pivotaltracker.com/story/show/163383200)

### CODE REVIEW

**Review by missvicki — APPROVED** [review]
Thank you for explaining this to me, I understand it a little bit better. Nice work you did here, I appreciate that you went ahead and implemented authors being able to see read stats of their articles.

**Review by fahdjamy — APPROVED** [review]

File: authors/apps/articles/views.py [hunk 1]
```diff
@@ -61,13 +63,17 @@ def get_object(self, slug):
...
-        return Response({
-            'errors': 'sorry article with that slug doesnot exist'
-        }, status=status.HTTP_404_NOT_FOUND)
+        if not article:
+            return Response({
+                'errors': 'sorry article with that slug doesnot exist'
+            }, status=status.HTTP_404_NOT_FOUND)
+        serialized_data = self.serializer_class(article,
+                                                context=context)
+        if not request.user.is_anonymous and\
```
Comment (karuhanga) [inline authors/apps/articles/views.py:74]: A few comments here would be helpful @codjoero

File: authors/apps/profiles/urls.py [hunk 2]
```diff
@@ -18,5 +19,8 @@
          name="following"),
     path("follow/followers", ProfileFollowersAPIView.as_view(),
          name="followers"),
-    path("", UserProfileListAPIView.as_view(), name="profile-list"),
+    path("", UserProfileListAPIView.as_view(),
+         name="profile-list"),
+    path('<username>/readstats', ReadStatsView.as_view(),
```
Comment (karuhanga) [inline authors/apps/profiles/urls.py]: Why do you need to specify a `username` if the user is already `logged in`, and from what I understand, a user can `only` get `their` `stats`?

File: authors/apps/articles/models.py [hunk 3]
```diff
@@ -60,6 +60,7 @@ class Article(models.Model):
     disliked_by = models.ManyToManyField(to=settings.AUTH_USER_MODEL,
                                          related_name='disliked_articles',
                                          related_query_name='disliked_article')
+    read_stats = models.IntegerField(default=0)
```
Comment (karuhanga) [inline authors/apps/articles/models.py]: I would like to understand the use of this field, and where exactly it is being used.

**Review by karuhanga — CHANGES_REQUESTED** [review]
Good work on this @codjoero.
I have left specific comments below.

File: authors/apps/articles/models.py (hunk 3 above)
Comment (codjoero) [reply authors/apps/articles/models.py]: @Karuhanga , thanks for the catch, had planned to use it to store the Article readstats before i changed to a MethodSerializer.

File: authors/apps/profiles/urls.py (hunk 2 above)
Comment (codjoero) [reply authors/apps/profiles/urls.py]: @Karuhanga As you might notice if you read the API spec, `GET /profiles/:username` would try to fetch you the profile of `<username>` hence the use of `GET /profiles/:username/readstats` to fetch the user's reading stats.

File: authors/apps/articles/views.py (hunk 1 above)
Comment (codjoero) [reply authors/apps/articles/views.py:74]: @Karuhanga thanks for the suggestion, I went ahead and added some comments where I thought the code was abit complex.

**Comment by pbkabali** [comment]
Great work here @codjoero. Your code is easy to follow and understand.

**Review by pbkabali — APPROVED** [review]

**Review by malep2007 — APPROVED** [review]

**Comment by malep2007** [comment]
@codjoero ,please resolve merge conflicts.

File: authors/apps/profiles/urls.py (hunk 2 above)
Comment (karuhanga) [reply authors/apps/profiles/urls.py]: True, I have noticed that. But in this case one cannot retrieve another user's reading stats the way they can retrieve their profile. In light of this therefore, I do not see the need to require the API consumer to supply a user for a user that is already logged in.

Comment (codjoero) [reply authors/apps/profiles/urls.py]: @Karuhanga , thanks for the feedback, the issue has been resolved

**Comment by malep2007** [comment]
@codjoero some merge conflits

**Pull request merged**